scala-protoc-plugins/scala-akka/AkkaGrpcServicePrinter.scala

```diff
@@ -12,91 +12,105 @@ final class AkkaGrpcServicePrinter(
 )(implicit descriptorImplicits: DescriptorImplicits) {
   import descriptorImplicits._
 
-  private val streamObserver = "_root_.io.grpc.stub.StreamObserver"
-  private val killSwitchName = s""""${service.getName}KillSwitch ${System.nanoTime()}""""
+  private val StreamObserver = "_root_.io.grpc.stub.StreamObserver"
 
-  private def observer(typeParam: String): String = s"$streamObserver[$typeParam]"
+  def printService(printer: FunctionalPrinter): Option[FunctionalPrinter] = {
+    val hasStreamingEndpoint: Boolean = service.methods.exists(_.isServerStreaming)
+
+    if (hasStreamingEndpoint) Some {
+      printer
+        .add(
+          "package " + service.getFile.scalaPackage.fullName,
+          "",
+          s"trait ${service.name}AkkaGrpc extends ${service.getName}Grpc.${service.getName} with AutoCloseable {",
+        )
+        .call(traitBody)
+        .add("}")
+    }
+    else None
+  }
+
+  private def responseType(method: MethodDescriptor): String = method.outputType.scalaType
+
+  private def observer(typeParam: String): String = s"$StreamObserver[$typeParam]"
 
   private def serviceMethodSignature(method: MethodDescriptor): PrinterEndo = { p =>
     method.streamType match {
       case StreamType.Unary => p
       case StreamType.ClientStreaming => p
       case StreamType.ServerStreaming =>
-        p.add(s"def ${method.name}(")
+        p
+          .add(s"def ${method.name}(")
           .indent
           .add(s"request: ${method.inputType.scalaType},")
-          .add(s"responseObserver: ${observer(method.outputType.scalaType)}")
+          .add(s"responseObserver: ${observer(responseType(method))}")
           .outdent
-          .add(s"): Unit = {")
+          .add("): Unit =")
           .indent
-          .add("if (closed.get()) {")
+          .add("synchronized {")
```

Comment:
This puts the whole source creation into a synchronized block. I do not think this is a big problem, but since we have no control on how long the computation takes until we have the source, I think it would be safer to restructure the synchronized block so it is not blocking on the underlying protected call for ${method.name}Source

Reply:
Good defensive thinking. Addressed in dce1585729543b992c2b37d71ec386b69385b4fe